[3기-C 장주영] JPA 게시판 구현 미션 제출합니다

cloudwi opened the pull request:
- cascadeType을 적용할때 고료해야하는 사항이 무엇인지 알고 싶습니다.
- controller -> service 단으로 데이터를 넘겨 줄때 Dto통으로 넘겨 주는 것과 Dto.get매서드를 해서 값을 넘겨주는 방식의 장단점이 궁금합니다.
- @PathVariable 와 @RequestParam을 사용하는 상황이 궁금합니다.
- CustomException 구현 시 실제 발생한 Exception도 넣어서 주는 것 이 좋습니까?

cloudwi committed:
요구사항에 등장하는 createdAt, createdBy를 자동적으로 입력 할 수 개발

cloudwi committed:
log를 모든 매서드마다 확인하기 위해서 aop를 적용하요 log를 찍음

cloudwi committed:
CustomException을 만들어서 에러 상황에 맞는 메세지를 전달 함

cloudwi committed:
Member entity와 MemberRepository를 구현함

cloudwi committed:
게시글의 요구사항 생성, 단건 조회, 전체 조회, 수정을 구현함

cloudwi committed:
rest api docs를 활용하여 백엔드 구현 정도를 보여줌.

리드미 파일로 프로젝트의 소개를 작성

cloudwi committed:
실행을 위한 파일 업로드

cloudwi committed:
실행을 위한 파일 업로드

cloudwi committed:
BaseTimeEntity 설정 시 데이터 베이스에 ms까지 저장되도록 세팅

- 하지만 사용자에게 보여줄 때는 사람이 인식 편한 초 단위 까지 세팅
- dto에서 포메터 사용하여 세팅

cloudwi committed:
log commit에 포함되지 않도록 설정

cloudwi committed:
ErrorResponse에 JsonFormat 적용

cloudwi committed:
Member에 CascadeType.All 적용

- member가 삭제된다면 연관관계된 Post도 함께 삭제 가능

cloudwi committed:
PostService getMember에 POST_NOT_FOUND -> MEMBER_NOT_FOUND 수정

cloudwi committed:
saveNotFoundMember라는 테스트 추가

- 게시글 생성 시 가입되지 않은 회원 아이디 입력 시 다음과 같은 오류 발생

cloudwi committed:
member 삭제 시에만 post 삭제 따라서 CascadeType.REMOVE로 변경

kimziou77 commented on src/main/java/com/example/springbootboardjpa/domain/member/entity/Member.java:
```diff
@@ -0,0 +1,42 @@
+    @Column
+    private String hobby;
+
+    @OneToMany(mappedBy = "member", cascade = CascadeType.REMOVE)
```
멤버가 삭제되면 모든 게시글이 사라지나 보군요!

cloudwi replied:
맞아요 리드미 파일에 추가를 했어야하는데 .... 실수했습니다.

kimziou77 commented on src/main/java/com/example/springbootboardjpa/domain/post/dto/response/PostFindAllResponseDto.java:
```diff
@@ -0,0 +1,21 @@
+
+    private long postId;
+    private String title;
+    @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")
+    private LocalDateTime createdAt;
```
저도 최근에 알았는데 시간 타입에 대한 부분도 굉장히 다양한 것들이 있더라구요.
LocalDateTime, Instant, ZonedDateTime, OffsetDateTime 등
그렇다고 여기가 문제가 되는 부분은 아닌데, 해당 차이점을 인지하면 더 다양한 상황에도 유연하게 대처할 수 있을 것 같습니다!

cloudwi replied:
너무 감사합니다.
상황에 맞게 시간 타입을 활용할 수 있을 것 같다는 생각이 들었습니다.
`Instant` 단일 진법으로 계산이 용이 하다는 특징이 있군요.
`LocalDateTime` 로컬 PC의 시간을 반영하는 군요 PC가 다르면 변화될 가능성도.... 있을 것 같습니다.
`ZoneDateTime` 서울의 시간으로 설정 하면 다음과 같아 지는 군요

kimziou77 replied:
자세한건 저도 더 봐야겠지만 타임존 여부랑 시간계산에서 차이가 있더라구요! 저의 경우 현재 하는 플젝이 해외사용자들도 발생해서 타임존이 포함되어있고 계산이 편한 Instant 데이터 타입을 사용하고 있습니다

kimziou77 commented on src/main/java/com/example/springbootboardjpa/domain/post/entity/Post.java:
```diff
@@ -0,0 +1,44 @@
+
+@Entity
+@Getter
+@NoArgsConstructor
```
데이터베이스 상에서는 post의 title이 not null로 생성이 되는데 NoArgsConstructor를 public으로 열어둔다면?!

cloudwi replied:
null 값을 가지고 있는 객체가 생성되고 저장될 가능성이 있네요...
따라서 @NoArgsConstructor(access = AccessLevel.PROTECTED)
로 변경해서 조금 더 안전하게 사용 가능할 것 생각합니다.

의견 감사합니다 !!!

kimziou77 replied:
반영된 리뷰는 밑에 Resolve 누르시면 닫힙니닷

kimziou77 commented on src/main/java/com/example/springbootboardjpa/domain/post/entity/Post.java:
```diff
@@ -0,0 +1,44 @@
+    @JoinColumn(name = "member_id", nullable = false)
+    private Member member;
+
+    public Post(String title, String content) {
+        this.title = title;
+        this.content = content;
+    }
```
title과 content가 모두 String 타입이어서 new Post("본문내용", "타이틀");
이런식으로 잘못쓰이면 컴파일러가 잡아주지 못한 에러가 날 수 있을 것 같아요!
받는 인자의 이름이 명확한 빌더 등을 활용해봐도 좋을 것 같습니다!

cloudwi replied:
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor
@Builder

Post class에 다음 3단 콤보를 사용해서 빌더를 사용해 보았습니다.
`= new ArrayList<>();` 같은 생성해주는 것은 @Builder.Default가 필요하다는 것두 알았습니다.

그리고 추가적인 질문으로 
`public void update(String title, String content) {
    this.title = title;
    this.content = content;
  }`

이런한 매서드도 리뷰 주신것과 같은 실수가 발생할 수 있다고 생각합니다. 이 매서드도 빌더 형식으로 작동하게 만드는게 좋을까요?

cloudwi replied:
`public void changeTitle(String title) {
    this.title = title;
  }

  public void changeContent(String content) {
    this.content = content;
  }`
  
  다음과 같이 매서드로 각각의 필드를 분리하는 .... setter같은 친구들을 두었습니다.

kimziou77 replied:
> 그리고 추가적인 질문으로
public void update(String title, String content) { this.title = title; this.content = content; }
이런한 매서드도 리뷰 주신것과 같은 실수가 발생할 수 있다고 생각합니다.

생성할때 member가 확정값으로 들어가도록 만들어 놓으면 update 사용할때 member가 안들어가있는 상황은 없을 것 같은데 어떤 부분에서 그렇게 생각하셨나요??

update에 할 수 있는 상황이 한 문서를 여럿이 수정할 수 있는 경우, 작성자만 변경이 가능한 경우 등등..이 있을 것 같은데
- 첫번째의 경우 createdBy에 첫 작성자가 기록되고, 누군가 수정을 했다면 modifiedBy에 기록하는 방법이 있을수 있고
- 두번째의 경우 update 하기 전에 작성자에 해당하는 member 만 변경할 수 있게 막을 수도 있을 것 같습니다

cloudwi replied:
게시글의 본문과 내용을 수정할 수 있는 매서드 update에서 다음과 같은 `update("본문내용", "타이틀");` 로 작성하는 오류가 발생할 수 있다는 질문 이었습니다.

따라서 각각의 필드의 update 매서드를 만들어서 해결해보고자 합니다.

kimziou77 commented on src/main/java/com/example/springbootboardjpa/domain/post/service/PostService.java:
```diff
@@ -0,0 +1,64 @@
+    public Post save(String title, String content, long memberId) {
+        Member member = getMember(memberId);
+
+        Post post = new Post(title, content);
+        post.setMember(member);
```
![image](https://user-images.githubusercontent.com/41179265/211895039-93753e8b-ca41-464e-b4cd-b7bf1834942a.png)
post에 있는 member는 반드시 존재해야 하는 값으로 보입니다.
post가 존재하기 위해서 title, content, member가 필수값이라면 setter로 member를 넣는것 보다는
생성자부터 member를 받아서 생성해주는 것이 더욱 좋아 보입니다. 현재는 실수로 setMember 안해주면 에러날 것 같아여

cloudwi replied:
`Post post = Post.builder()
        .title(title)
        .content(content)
        .member(member)
        .build();`

다음과 같이 수정 하였습니다.
하지만 post에 회원을 수정할 비지니스로직이 발생할 경우 

`public void setMember(Member member) {
    if (this.member != null) {
      this.member.getPosts().remove(this);
    }
    this.member = member;
    member.getPosts().add(this);
    this.setCreatedBy(member.getId());
  }`

위 매서드는 필요하다고 생각이 됩니다.

kimziou77 replied:
setMember 가 꼭 필요한 비지니스로직이 발생할 상황이 온다면 그때 추가하면 좋을것 같아요

수정하는 값이 많아지는 경우 PostHistory 같은 클래스에 변경되는 값들을 쌓아보는 것도 좋을 것 같아여

이런 느낌 이랄까요..!
![image](https://user-images.githubusercontent.com/41179265/212174327-71c3ef19-d6c3-49d3-a11f-436ca62214ef.png)

cloudwi replied:
히스토리 클래스로 기록을 남긴다. 생각해두겠습니다.
하지만 이번 프로그램에서는 큰 필요성을 느끼지 못해서 추가는 안하겠습니다. 하지만 여유가 된다만 만들어보고 싶네요 !!! 
추구 추가하겠습니다.

kimziou77 commented on src/main/java/com/example/springbootboardjpa/domain/post/service/PostService.java:
```diff
@@ -0,0 +1,64 @@
+    @Transactional
+    public Post update(long id, String title, String content, long memberId) {
+        Post post = getPost(id);
+        idMatching(post.getMember().getId(), memberId);
```
post에게 memberId를 전달해서 해당 멤버가 작성자인지 물어보는 메서드가 있어도 좋을 것 같습니다.
post가 가지고 있는 값을 너무 쉽게 강탈해 가는 것 같습니다.(심지어 post를 통해서 member아이디 뺏음)
post랑 member 프라이버시도 존중해주세요!

cloudwi replied:
반대는 어떻게 생각하시나요?
member 한테 post.getMemberId해서 주고 확인 해달라구 하는 것 입니다.
먼가 memberId는 member에서 검증해야할 것 같은 생각이.... 있습니다....

`class Member { 
...
public void idMatching(Long id) {
    if (this.id != id) {
      throw new CustomException(ErrorCode.NOT_HAVA_PERMISSION, new IllegalArgumentException());
    }
  }
}`

kimziou77 replied:
코드 블럭은 ` ``` ` 로 작성할 수 있습니닷

kimziou77 replied:
post에 작성자 정보 있고, update 하려는 memberId는 받고 있어서
post거 아이디도 뺏을 필요가 크게 없다고 생각합니다. `post.isOwner(memberId);` 정도는 어떠신가요?

cloudwi replied:
```
class Member
```

cloudwi replied:
너무 좋네요 ```Post```가 갑자기 불쌍해 보이기 시작했습니다.🥹

```
public void isOwner(long memberId) {
    this.member.idMatching(memberId);
  }
```

다음과 같이 수정 하였습니다!

cloudwi replied:
누구에게도 그 값을 뺏지 않고 값을 줘서 비교해줘 라고는 말을 하고 있습니다.

kimziou77 commented on src/main/resources/application.yml:
```diff
@@ -0,0 +1,52 @@
+      production:
+        - prod
+
+  datasource:
+    driver-class-name: org.h2.Driver
+    url: jdbc:h2:file:~/springBoardJpa
+    username: sa
+    password:
```
데이터베이스는 3개 환경 모두 동일한 파일을 사용하는건가용?

cloudwi replied:
원래라면 다르게 해야겠다는 생각은 가지고 있지만 현재는 동일한 데이터 베이스를 사용하고 있습니다....

kimziou77 replied:
환경 나누셨으니 디비도 나눠주시면 좋을 것 같습니다 ㅎㅎ

cloudwi replied:
넵!! 데이터 베이스를 분리 하였습니다.
local에서는 ~:mem:local 사용하지만 ....

kimziou77 commented on src/main/java/com/example/springbootboardjpa/global/exception/ErrorCode.java:
```diff
@@ -0,0 +1,24 @@
+    INVALID_AUTH_TOKEN(HttpStatus.UNAUTHORIZED, "잘못된 토큰입니다."),
+    POST_NOT_FOUND(HttpStatus.NOT_FOUND, "게시글을 찾지 못했습니다."),
+    MEMBER_NOT_FOUND(HttpStatus.NOT_FOUND, "회원을 찾지 못했습니다."),
+    NOT_HAVA_PERMISSION(HttpStatus.NOT_ACCEPTABLE, "권한이 없습니다.");
```
NOT_ACCEPTABLE 은 무엇은 하는 상태코드인가요? 더 적절한 상태코드가 있을 것 같습니다

cloudwi replied:
[403 Forbidden]
클라이언트가 권한이 없기 때문에 작업을 진행할 수 없는 경우 금지됨

다른 사람이 작성한 글을 수정하려고 하는 것을 금지시키는 에러 코드로 수정 하겠습니다.

kimziou77 commented:
**cascadeType을 적용할때 고료해야하는 사항이 무엇인지 알고 싶습니다.**
```왜 cascade를 적용해야 했는지 대답하실 수 있다면 적용해보셔도 좋다고 생각합니다.```

**controller -> service 단으로 데이터를 넘겨 줄때 Dto통으로 넘겨 주는 것과 
Dto.get매서드를 해서 값을 넘겨주는 방식의 장단점이 궁금합니다.**
```
음 DTO를 통으로 넘겨주는 상황을 기준으로 생각을 해보면
- 장) request로 받는 값의 데이터 변경이 일어나도 메서드 시그니처를 변경하지 않아도 된다
- 단) 해당 서비스는 해당 DTO를 받아야만 동작할 수 있다?
아직 해당 부분으로 크게 머 문제가 생긴 적은 없어서 저정도 장단점 있지 않을까 싶네요
```

**@PathVariable 와 @RequestParam을 사용하는 상황이 궁금합니다.**
```
음 @PathVariable같은 경우는 폴더구조와 비슷하다고 생각하는데
/XX초등학교/6학년/3반/15번/김수빈 -> 이런 계층적으로 자원의 관계를 알 수 있는 느낌이라면

@RequestParam같은 경우에는 자원에 대한 질의를 할때 사용하는 것 같습니다.
/XX초등학교?학년=6&반=3&출석번호=15 -> 응답: 김수빈
```

 **CustomException 구현 시 실제 발생한 Exception도 넣어서 주는 것 이 좋습니까?**
```
Throwable 도 받을 수 있게 메서드를 만들면 좋을 것 같습니다.
주영님 코드를 조금 빌려보자면

public CustomException(ErrorCode errorCode, Throwable cause) {
    super(message, cause);
    this.errorCode = errorCode;
}

대충 이런 느낌
```

cloudwi commented:
감사 합니다 반영해서 commit 하겠습니다!

cloudwi committed:
PR 리뷰 반영

kimziou77 commented on src/main/resources/application.yml:
```diff
@@ -9,7 +9,7 @@ spring:
   datasource:
     driver-class-name: org.h2.Driver
-    url: jdbc:h2:file:~/springBoardJpa
+    url: jdbc:h2:mem:test
```
ㅋㅋㅋㅋ 왜 안나누고 mem으로 바꾸신거죠... 심지어 이름이 test..
근데 local 환경은 따로 정의 안되어있나요? dev랑 prod만 사용하시는건가요?

cloudwi replied:
편하게 작업한다는 생각에 인메모리 형식으로 전부 사용했습니다 .....;;;
변경하였습니다 !

cloudwi replied:
로컬 추가했습니다.

```
spring:
  profiles:
    active: dev
    group:
      development:
        - dev
      production:
        - prod
      local:
        - local
```

lleellee0 commented on src/main/resources/application.yml:
```diff
@@ -0,0 +1,52 @@
+        show_sql: true
+    database: h2
+    show-sql: true
+
+---
+
+spring:
+  config:
+    activate:
+      on-profile: dev
+
+
+logging:
+  file:
+    path: ./logs/app-dev
+
+---
+
+spring:
+  config:
+    activate:
+      on-profile: prod
+
+logging:
+  file:
+    path: ./logs/app-release
```
Profile을 쪼개보시면 좋을 것 같아요.

cloudwi replied:
혹쉬 yml 파일로 쪼개는걸 말씀하시나요 
--- 으로 쪼개는게 아닌가요?
😅

lleellee0 commented on src/main/java/com/example/springbootboardjpa/global/aop/CommonPointcut.java:
```diff
@@ -0,0 +1,10 @@
+
+public class CommonPointcut {
+
+  @Pointcut("execution(public * com.example.springbootboardjpa..*.*(..))")
+  public void packagePublicMethodPointcut() {
+  }
```
이거 AOP 걸어서 뭘 하려는거죵

lleellee0 replied:
아 밑에 있네요 ㅋ

cloudwi replied:
아하 모든 매서드가 실행되고 종료되는 시점에 로그를 남기고 싶었습니다. 개발 환경에서 이슈를 찾기 좋아서 사용하고 있습니다!!

lleellee0 commented on src/main/java/com/example/springbootboardjpa/domain/member/entity/Member.java:
```diff
@@ -0,0 +1,54 @@
+  @Column
+  private String hobby;
+
+  @Builder.Default
+  @OneToMany(mappedBy = "member", cascade = CascadeType.REMOVE)
+  private List<Post> posts = new ArrayList<>();
```
CascadeType.REMOVE 로 설정하신건 의도 하신 부분인거죠?

cloudwi replied:
넵 회원이 삭제되면 
그 회원이 작성한 게시글 전부 삭제되도록 비지니스를 생각해 보았습니다. 추가적으로 Cascade 연습도 해볼겸사겸사 였습니다....😱

lleellee0 commented on src/main/java/com/example/springbootboardjpa/global/exception/ErrorCode.java:
```diff
@@ -0,0 +1,24 @@
+  INVALID_AUTH_TOKEN(HttpStatus.UNAUTHORIZED, "잘못된 토큰입니다."),
+  POST_NOT_FOUND(HttpStatus.NOT_FOUND, "게시글을 찾지 못했습니다."),
+  MEMBER_NOT_FOUND(HttpStatus.NOT_FOUND, "회원을 찾지 못했습니다."),
+  NOT_HAVA_PERMISSION(HttpStatus.FORBIDDEN, "권한이 없습니다.");
```
`NOT_HAVA_PERMISSION` HAVA가 무슨 의미죠? 혹시 HAVE인건지..

cloudwi replied:
분명 번역기를 돌려서 만들었던 걸루 기억하지만... 디버그 하보니 저도 이해가 되지 않는 문구 이군요.....
아래와 같이 변경하였습니다.
권한없음을 의미하도록 하였습니다.!🌪️
`NOT_PERMISSION`

lleellee0 commented on src/main/java/com/example/springbootboardjpa/domain/member/entity/Member.java:
```diff
@@ -0,0 +1,54 @@
+    this.hobby = hobby;
+  }
+
+  public void idMatching(Long id) {
+    if (this.id != id) {
+      throw new CustomException(ErrorCode.NOT_HAVA_PERMISSION);
+    }
+  }
```
idMatching 라는 이름이 메서드의 이름으로 적합해보이시나요?

cloudwi replied:
적합 하지 않은것 같습니다...👀
따라서 공부 해 보았습니다.

https://tecoble.techcourse.co.kr/post/2020-04-26-Method-Naming/

위의 글을 보고 

isId 로 수정해는게 좋을까 라는 생각을 하였습니다.
<img width="596" alt="image" src="https://user-images.githubusercontent.com/86584887/212513932-10144052-ad3b-4d22-84a9-005fb3be4b31.png">

lleellee0 commented on src/main/java/com/example/springbootboardjpa/domain/post/contorller/PostController.java:
```diff
@@ -0,0 +1,103 @@
+    Page<Post> posts = postService.findAll(pageRequest);
+
+    posts.forEach(post -> {
+      PostFindAllResponseDto postFindAllResponseDto = new PostFindAllResponseDto(
+          post.getId(),
+          post.getTitle(),
+          post.getCreatedAt(),
+          post.getCreatedBy()
+      );
```
이런 코드는 PostFindAllResponseDto 안으로 들어가면 어떨까요?
Post를 파라미터로 받는 정적 팩토리 메서드가 있으면 이런 코드가 반복되는걸 막으면서도 가독성을 높일 수 있을 것 같습니다.

cloudwi replied:
```
public PostFindAllResponseDto(Post post) {
        this.postId = post.getId();
        this.title = post.getTitle();
        this.createdBy = post.getCreatedBy();
        this.createdAt = post.getCreatedAt();
    }
```

이렇게 하니까 확실히 코드 가독성이 엄청 높아집니다!!!

lleellee0 commented on src/main/java/com/example/springbootboardjpa/domain/post/contorller/PostController.java:
```diff
@@ -0,0 +1,103 @@
+  @GetMapping("/{postId}")
+  public ResponseEntity<PostFindByIdResponseDto> findById(@PathVariable long postId) {
+    Post post = postService.findById(postId);
+    PostFindByIdResponseDto postFindByIdResponseDto = new PostFindByIdResponseDto(
+        post.getId(),
+        post.getTitle(),
+        post.getContent(),
+        post.getCreatedBy(),
+        post.getCreatedAt()
+    );
```
https://github.com/prgrms-be-devcourse/springboot-board-jpa/pull/211/files#r1068907190

마찬가지

cloudwi replied:
수정하였습니다!

lleellee0 commented on src/main/java/com/example/springbootboardjpa/domain/post/contorller/PostController.java:
```diff
@@ -0,0 +1,103 @@
+        postSaveRequestDto.getMemberId()
+    );
+
+    PostSaveResponseDto postSaveResponseDto = new PostSaveResponseDto(
+        post.getId(),
+        post.getTitle(),
+        post.getContent(),
+        post.getCreatedAt(),
+        post.getCreatedBy()
+    );
```
https://github.com/prgrms-be-devcourse/springboot-board-jpa/pull/211/files#r1068907190

마찬가지

cloudwi replied:
수정하였습니다.🏵️

lleellee0 commented on src/main/java/com/example/springbootboardjpa/domain/post/contorller/PostController.java:
```diff
@@ -0,0 +1,103 @@
+        postUpdateRequestDto.getMemberId()
+    );
+
+    PostUpdateResponseDto postUpdateResponseDto = new PostUpdateResponseDto(
+        post.getId(),
+        post.getTitle(),
+        post.getContent(),
+        post.getCreatedBy(),
+        post.getCreatedAt()
+    );
```
https://github.com/prgrms-be-devcourse/springboot-board-jpa/pull/211/files#r1068907190

마찬가지

cloudwi replied:
수정하였습니다.🧗‍♀️

lleellee0 commented on src/main/java/com/example/springbootboardjpa/domain/post/service/PostService.java:
```diff
@@ -0,0 +1,68 @@
+    return getPost(id);
+  }
+
+  @Transactional
+  public Post update(long id, String title, String content, long memberId) {
+    Post post = getPost(id);
```
update 메서드가 너무 Post의 필드 변화에 영향을 많이 받고 있습니다.
이건 캡슐화가 깨진거에요.
다른 방법이 없을까요?

cloudwi replied:
아래와 같이 수정하였습니다!
Jpa에서 id가 같은 값은 같은 객체로 보기때문에 save를 할때 같은 id를 가지고 있으면 덮어쓰기가 처럼 수정된다는 것을 알게 되었습니다.

```merge()```

하지만 save()를 2번 하면 Post가 상속하고 있는 BaseTimeEntity에 created_at에 null 값이 들어가는 것을 확인했습니다.... 
[위 내용을 찾은 블로그](https://jaime-note.tistory.com/65)

따라서 
```
@CreatedDate
  @LastModifiedDate
  private LocalDateTime createdAt;
```

로 변경하였습니다. 수정되면 게시글은 새롭게 탄생하는 것이나 다름 없다고 비지니스 로직을 생각하였습니다.

```
@Transactional
  public Post update(long postId, String title, String content, long memberId) {
    Post post = getPost(postId);
    post.isOwner(memberId);

    Member member = getMember(memberId);

    Post updatePost = Post.builder()
        .id(postId)
        .title(title)
        .content(content)
        .member(member)
        .build();

    return postRepository.save(updatePost);
  }
```_

lleellee0 commented on src/main/java/com/example/springbootboardjpa/domain/post/service/PostService.java:
```diff
@@ -0,0 +1,68 @@
+    return post;
+  }
+
+  private Post getPost(long id) {
+    return postRepository.findById(id).orElseThrow(() ->
+        new CustomException(ErrorCode.POST_NOT_FOUND));
+  }
+
+  private Member getMember(long memberId) {
+    return memberRepository.findById(memberId).orElseThrow(() ->
+        new CustomException(ErrorCode.MEMBER_NOT_FOUND));
+  }
```
CustomException 만을 정의해서 사용하고 있는데, 예외의 타입을 세분화할 필요성은 없을까요?

cloudwi replied:
CustomException안에 ErrorCode만으로는 부족합니까?

lleellee0 commented on src/main/java/com/example/springbootboardjpa/global/aop/LogAspect.java:
```diff
@@ -0,0 +1,36 @@
+  public Object logging(ProceedingJoinPoint pjp) throws Throwable {
+    long beforeTimeMillis = System.nanoTime();
+
+    log.info("실행 시작 : "
+        + pjp.getSignature().getDeclaringTypeName() + "."
+        + pjp.getSignature().getName());
+
+    Object result = pjp.proceed();
+
+    long afterTimeMillis = System.nanoTime() - beforeTimeMillis;
+
+    log.info("실행 완료 : "
+        + afterTimeMillis + "ns 소요 : "
+        + pjp.getSignature().getDeclaringTypeName()
+        + " : "
+        + pjp.getSignature().getName());
```
이거 info 레벨로 계속 로그를 찍는게 적절한지는 실제 서비스라면 검토가 필요해보입니다.

cloudwi replied:
문제를 선별하기 위해서 로그를 찍는 것이기 때문에 디버그 로그가 적절하다고 생각이 바뀌었습니다....

lleellee0 commented on src/main/java/com/example/springbootboardjpa/global/exception/ErrorResponse.java:
```diff
@@ -0,0 +1,48 @@
+
+  @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")
+  private final LocalDateTime timestamp = LocalDateTime.now();
+  private final Integer status;
```
에러 응답 Body에 status code 를 넣지 않는걸 추천드립니다.
클라이언트가 헤더에 담긴 status code가 아니라 바디에 담긴걸 사용할 우려가 있어요.

cloudwi replied:
클라이언트가 헤더랑 바디에 있는 데이터를 보고 헷갈릴 수 있다고 생각하게 되었습니다.!! 감사합니다.

다음과 같이 수정 하였습니다.
```
package com.example.springbootboardjpa.global.exception;


import com.fasterxml.jackson.annotation.JsonFormat;
import lombok.Builder;
import lombok.Getter;
import org.springframework.http.ResponseEntity;

import java.time.LocalDateTime;

@Getter
@Builder
public class ErrorResponse {

  @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")
  private final LocalDateTime timestamp = LocalDateTime.now();
  private final Integer status;
  private final String error;
  private final String code;
  private final String message;

  public static ResponseEntity<ErrorResponse> toResponseEntity(ErrorCode errorCode) {

    return ResponseEntity
        .status(errorCode.getHttpStatus())
        .body(
            ErrorResponse.builder()
                .code(errorCode.name())
                .message(errorCode.getMessage())
                .build());
  }

  public static ResponseEntity<Object> toObject(CustomException customException) {
    ErrorCode errorCode = customException.getErrorCode();

    return ResponseEntity
        .status(errorCode.getHttpStatus())
        .body(
            ErrorResponse.builder()
                .code(errorCode.name())
                .message(customException.getMessage())
                .build());
  }
}
```

lleellee0 commented on src/test/java/com/example/springbootboardjpa/domain/post/contorller/PostControllerTest.java:
```diff
@@ -0,0 +1,260 @@
+import static org.springframework.restdocs.request.RequestDocumentation.*;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;
+
+@SpringBootTest
+@AutoConfigureRestDocs
+@AutoConfigureMockMvc
+class PostControllerTest {
```
컨트롤러에 대한 통합 테스트는 잘 작성해주셨네요.
각 계층별로 단위 테스트도 작성해보시면 좋을 것 같습니다. ^^

cloudwi replied:
moktio를 사용해서 postServiceTest를 작성하였습니다 !!! 
감사합니다 !

lleellee0 commented:
주영님 수고하셨습니다.

> cascadeType을 적용할때 고료해야하는 사항이 무엇인지 알고 싶습니다.

리뷰에도 남겨놨는데, cascade 되는게 비즈니스적으로 의도된 바인지 따지는게 첫번째로 중요합니다. 그거 외에도 cascade 되면서 N+1 문제가 발생할 수도 있는데요, 그런 부분까지 고려되면 크게 문제 없습니다.

> controller -> service 단으로 데이터를 넘겨 줄때 Dto통으로 넘겨 주는 것과 Dto.get매서드를 해서 값을 넘겨주는 방식의 장단점이 궁금합니다.

장단점은 주영님도 느끼실 수 있지 않나요?
DTO를 서비스까지 가져오는거랑 컨트롤러까지만 가져오는거랑 어떤 차이가 있는지 생각해보시기 바랍니다.
더 궁금하면 아래 링크도 읽어보세요.
https://github.com/HomoEfficio/dev-tips/blob/master/DTO-DomainObject-Converter.md

> @PathVariable 와 @RequestParam을 사용하는 상황이 궁금합니다.

PathVariable -> 경로의 일부이기 때문에 자원을 표현할 때(주로 자원의 식별자)
RequestParam -> 자원의 식별자 보다는 주로 검색 조건 등을 지정할 때

이렇게 나눠서 사용합니다.

> CustomException 구현 시 실제 발생한 Exception도 넣어서 주는 것 이 좋습니까?

이건 필요하다면 넣어주면 되는데요, 주영님 코드에서 필요하다고 생각하시나요?
그렇다면 넣으면 됩니다.
그게 아니라면 넣지 않으셔도 됩니다.

cloudwi committed:
controller -> service
dto 매개변수로 사용 가능하도록 만듬

- dto를 서비스에서 알게되는 것을 피하기 위해서 값들을 뽑아서 사용하였지만 개발의 편의성과 생산성을 위해서 dto를 매개변수로 사용하도록 수정함

mokito 테스트 코드 작성

- 유닛 테스트를 만들기위해서 레포지토리등을 mock해줌
